### PR TITLE
Websocket + HA refactoring

### PR DESCRIPTION
**Pull request by safchain**

### CODE REVIEW

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
rerun tests

File: http/wsclient.go [hunk 1]
```diff
@@ -302,71 +308,88 @@ func (c *WSAsyncClient) Connect() {
...
+	c := &WSClient{
+		WSConn:     wsconn,
+		Path:       path,
+		AuthClient: authClient,
+	}
+	wsconn.wsSpeaker = c
+	return c
+}
+
+// NewWSAsyncClientFromConnection creates a client based on the configuration
```
Comment (lebauce) [inline http/wsclient.go]: Wrong comment

File: http/wsclient.go [hunk 2]
```diff
@@ -302,71 +308,88 @@ func (c *WSAsyncClient) Connect() {
...
+		pingTicker: &time.Ticker{},
 	}
 	c.connected.Store(false)
 	c.running.Store(true)
 	return c
 }
 
-// NewWSAsyncClientFromConnection creates a client from an existing connection
-func NewWSAsyncClientFromConnection(host string, clientType common.ServiceType, conn *websocket.Conn) *WSAsyncClient {
+// NewwsIncomingClient returns a client with a new connection
```
Comment (lebauce) [inline http/wsclient.go]: Wrong comment

File: http/pool.go [hunk 3]
```diff
@@ -32,220 +32,295 @@ import (
 	"github.com/skydive-project/skydive/config"
 )
 
-// WSClientPool represents a pool of WebSocket clients
-type WSClientPool struct {
+type WSSpeakerPool interface {
+	AddClient(c WSSpeaker) error
+	AddEventHandler(h WSSpeakerEventHandler)
+	GetClients() []WSSpeaker
+	GetConnectedClient() WSSpeaker
```
Comment (lebauce) [inline http/pool.go]: PickConnectedClient ?

Something that clearly states that a client is randomly chosen

File: http/pool.go [hunk 4]
```diff
@@ -32,220 +32,295 @@ import (
...
+	eventHandlersLock sync.RWMutex
+	clients           []WSSpeaker
+}
+
+// WSSpeakerPool represents a pool of WebSocket clients
+type WSClientPool struct {
+	*WSPool
+}
+
+// wsIncomerPool represents a pool of WebSocket conections
```
Comment (lebauce) [inline http/pool.go]: represents a pool of incoming WebSocket connections

File: http/pool.go [hunk 5]
```diff
@@ -32,220 +32,295 @@ import (
...
+	eventBuffer       []WSMessage
+	eventBufferLock   sync.RWMutex
+	wg                sync.WaitGroup
+	running           atomic.Value
+	eventHandlers     []WSSpeakerEventHandler
+	eventHandlersLock sync.RWMutex
+	clients           []WSSpeaker
+}
+
+// WSSpeakerPool represents a pool of WebSocket clients
```
Comment (lebauce) [inline http/pool.go]: WSClientPool represents a pool of outgoing WebSocket clients

File: http/wsclient.go [hunk 6]
```diff
@@ -302,71 +308,88 @@ func (c *WSAsyncClient) Connect() {
...
+	return c
+}
+
+// NewWSAsyncClientFromConnection creates a client based on the configuration
+func NewWSClientFromConfig(clientType common.ServiceType, addr string, port int, path string, authClient *AuthenticationClient) *WSClient {
+	host := config.GetConfig().GetString("host_id")
+	return NewWSClient(host, clientType, addr, port, path, authClient)
+}
+
+// newIncomingWSConn is called by the server for incoming connections
```
Comment (lebauce) [inline http/wsclient.go]: Change comment

**Comment by lebauce** [comment]
s/king/kind in the commit message

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
rerun unit-tests

**Comment by safchain** [comment]
rerun tests

**Comment by nplanel** [comment]
run skydive-scale-tests

**Comment by nplanel** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
run skydive-scale-tests

**Comment by lebauce** [comment]
run skydive-scale-tests

**Comment by lebauce** [comment]
run skydive-coverage

**Comment by lebauce** [comment]
run scale-tests

**Comment by safchain** [comment]
run scale-tests

**Comment by lebauce** [comment]
run scale-tests

**Pull request merged**